packages/eslint-plugin/src/rules/max-params.ts

```diff
@@ -0,0 +1,91 @@
+import { AST_NODE_TYPES, type TSESTree } from '@typescript-eslint/utils';
+
+import * as util from '../util';
+import { getESLintCoreRule } from '../util/getESLintCoreRule';
+
+type FunctionLike =
+  | TSESTree.FunctionDeclaration
+  | TSESTree.FunctionExpression
+  | TSESTree.ArrowFunctionExpression;
+
+type FunctionRuleListener<T extends FunctionLike> = (node: T) => void;
+
+const baseRule = getESLintCoreRule('max-params');
+
+export type Options = util.InferOptionsTypeFromRule<typeof baseRule>;
+export type MessageIds = util.InferMessageIdsTypeFromRule<typeof baseRule>;
+
+export default util.createRule<Options, MessageIds>({
+  name: 'max-params',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description:
+        'Enforce a maximum number of parameters in function definitions',
+      extendsBaseRule: true,
+    },
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          maximum: {
+            type: 'integer',
+            minimum: 0,
+          },
+          max: {
+            type: 'integer',
+            minimum: 0,
+          },
+          countVoidThis: {
+            type: 'boolean',
+          },
+        },
+        additionalProperties: false,
+      },
+    ],
+    messages: baseRule.meta.messages,
+  },
+  defaultOptions: [{ max: 3, countVoidThis: false }],
+
+  create(context, [{ countVoidThis }]) {
+    const baseRules = baseRule.create(context);
+
+    if (countVoidThis === true) {
+      return baseRules;
+    }
+
+    const removeVoidThisParam = <T extends FunctionLike>(node: T): T => {
+      if (node.params.length === 0) {
+        return node;
+      }
+
+      const params =
+        node.params[0] &&
+        node.params[0].type === AST_NODE_TYPES.Identifier &&
```

Comment:
[Nit] If `node.params.length` isn't `0` then we're guaranteed `node.params[0]` exists:

```suggestion
        node.params[0].type === AST_NODE_TYPES.Identifier &&
```

Reply:
...come to think of it, we can get rid of the new object altogether by also running `return node;` if we don't need to slice/shift the params. I'll quick edit that in.

Reply:
Yup, good catch
I want to blame copilot for this, but I don't remember whose fault it is 🫣

Reply:
😄 if you have to ask 😛 

(I also missed this until just now!)